File: C++/BellmanFord.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;
typedef long double ld;
#define pb push_back

struct Edge {
    int start, dest; ll weight;
};

const int MaxN = 1e5+10, MaxM = 1e5+10;
ll INF = 1e18;
Edge edges[MaxM];
ll dist[MaxN];
int n, m;

bool BF(int start) {
    for (int i=0; i<n; i++) dist[i] = INF;
    dist[start] = 0;
    for (int i=0; i<n-1; i++) {
        for (int j=0; j<m; j++) {
            int from = edges[j].start;
            int to = edges[j].dest;
            ll weight = edges[j].weight;
            dist[to] = min(dist[to], dist[from]+weight);
        }
    }
    
    // check for negative cycles
    for (int j=0; j<m; j++) {
        int from = edges[j].start;
        int to = edges[j].dest;
        ll weight = edges[j].weight;
        if (dist[from]+weight < dist[to]) return false;	// negative cycle
    }
    return true;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> n >> m;
    for (int i=0; i<m; i++) {
        cin >> edges[i].start >> edges[i].dest >> edges[i].weight;
        edges[i].start--; edges[i].dest--;
    }
    
    int start;
    cin >> start;
    start--;
    BF(start);
}

File: Java/BellmanFord.java

import java.util.*;
import java.io.*;

public class BellmanFord {

	static int n, m;
	static long INF = (long)(1e18);
	static Edge[] edges;
	static long[] dist;
	
	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		//BufferedReader in = new BufferedReader(new FileReader("BellmanFord"));

		StringTokenizer st = new StringTokenizer(in.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		dist = new long[n];
		edges = new Edge[m]; // 2*m if undirected
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(in.readLine());
			int one = Integer.parseInt(st.nextToken()) - 1;
			int two = Integer.parseInt(st.nextToken()) - 1;
			long weight = Long.parseLong(st.nextToken());
			edges[i] = new Edge(one, two, weight);
			// edges[i+m] = new Edge(two, one, weight);		if undirected
		}
		
		int start = Integer.parseInt(in.readLine())-1;
		bellmanFord(start);
		
	}
	
	public static boolean bellmanFord(int start) {
		Arrays.fill(dist, INF);
		dist[start] = 0;
		for (int i = 0; i < n-1; i++) {
			for (int j = 0; j < m; j++) {
				int from = edges[j].start;
				int to = edges[j].dest;
				long weight = edges[j].weight;
				dist[to] = Math.min(dist[to], dist[from] + weight);
			}
		}
		
		// check for negative cycles
		for (int j = 0; j < m; j++) {
			int from = edges[j].start;
			int to = edges[j].dest;
			long weight = edges[j].weight;
			if (dist[from] + weight < dist[to]) return false;	// negative cycle
		}
		return true;
	}
	
	static class Edge {
		int start, dest;
		long weight;
		Edge (int start, int dest, long weight) {
			this.start = start;
			this.dest = dest;
			this.weight = weight;
		}
	}
}